### PR TITLE
Add composable functions to Composer

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

This PR adds `functions` to `Composer` so that we can compose a `Composer` with a sequence of `Functions`. One use case is that we can reuse the `Mask` and `Additive` functions to compose a `MaskAdditive` Composer.

* Add the `mart.attack.composer.Function` base class and three children classes `Additive`, `Mask` and `Overlay`.
* Add `Composer(functions:dict)` so it is configurable in Hydra.
* Add `Function(order:int)` so we can sort the sequence of `Function`s in `Composer`.
* Update existing configs.
* Update tests.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [ ] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70% (21 sec/epoch).
- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70% (14 sec/epoch).

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: mart/attack/composer.py [hunk 1]
```diff
@@ -7,12 +7,37 @@
...
+
+    @abc.abstractmethod
+    def forward(self, perturbation, input, target) -> None:
+        """Returns the modified perturbation, modified input and target, so we can chain Functions
+        in a Composer."""
+        pass
+
+
+class Composer:
+    def __init__(self, functions: dict[str, Function]) -> None:
```
Comment (dxoigmn) [inline mart/attack/composer.py:34]: I would just do **names_and_functions, so you don't need to pass a dict. That is,

```python
Composer(mask=Mask(order=0), additive=Additive(order=1))
```

File: mart/attack/composer.py [hunk 2]
```diff
@@ -38,43 +63,55 @@ def __call__(
         else:
             raise NotImplementedError
 
-    @abc.abstractmethod
     def compose(
```
Comment (dxoigmn) [inline mart/attack/composer.py:66]: Should this just be private now?

File: mart/attack/composer.py [hunk 3]
```diff
@@ -7,12 +7,37 @@
...
+        """A stackable function for Composer.
+
+        Args:
+            order (int, optional): The priority number. A smaller number makes a function run earlier than others in a sequence. Defaults to 0.
+        """
+        super().__init__(*args, **kwargs)
+        self.order = order
+
+    @abc.abstractmethod
+    def forward(self, perturbation, input, target) -> None:
```
Comment (dxoigmn) [inline mart/attack/composer.py:27]: Return type annotation should be `tuple` of whatever types `perturbation, input, target` are.

**Review by dxoigmn — CHANGES_REQUESTED** [review]
See comments and suggestions.

File: mart/attack/composer.py (hunk 1 above)
Comment (mzweilin) [reply mart/attack/composer.py:34]: Yes, but leaving `functions` here makes it much easier to do modality_dispatch later.

File: mart/attack/composer.py (hunk 2 above)
Comment (mzweilin) [reply mart/attack/composer.py]: Yes. Fixed in b58b1077ccd4b6036420f5aab37409003b0f1906

File: mart/attack/composer.py (hunk 3 above)
Comment (mzweilin) [reply mart/attack/composer.py]: Yes. Added in 7eccbdd7eb4034203997b400bbfc86ba3603321d

**Review by dxoigmn — APPROVED** [review]
LGTM!

**Pull request merged**